Add `core::stream::Stream`

yoshuawuyts opened the pull request:
[[Tracking issue: #79024](https://github.com/rust-lang/rust/issues/79024)]

This patch adds the `core::stream` submodule and implements `core::stream::Stream` in accordance with [RFC2996](https://github.com/rust-lang/rfcs/pull/2996). The RFC hasn't been merged yet, but as requested by the libs team in https://github.com/rust-lang/rfcs/pull/2996#issuecomment-725696389 I'm filing this PR to get the ball rolling.

## Documentatation

The docs in this PR have been adapted from [`std::iter`](https://doc.rust-lang.org/std/iter/index.html), [`async_std::stream`](https://docs.rs/async-std/1.7.0/async_std/stream/index.html), and [`futures::stream::Stream`](https://docs.rs/futures/0.3.8/futures/stream/trait.Stream.html). Once this PR lands my plan is to follow this up with PRs to add helper methods such as `stream::repeat` which can be used to document more of the concepts that are currently missing. That will allow us to cover concepts such as "infinite streams" and "laziness" in more depth.

## Feature gate

The feature gate for `Stream` is `stream_trait`. This matches the `#[lang = "future_trait"]` attribute name. The intention is that only the APIs defined in RFC2996 will use this feature gate, with future additions such as `stream::repeat` using their own feature gates. This is so we can ensure a smooth path towards stabilizing the `Stream` trait without needing to stabilize all the APIs in `core::stream` at once. But also don't start expanding the API until _after_ stabilization, as was the case with `std::future`.

__edit:__ the feature gate has been changed to `async_stream` to match the feature gate proposed in the RFC.

## Conclusion

This PR introduces `core::stream::{Stream, Next}` and re-exports it from `std` as `std::stream::{Stream, Next}`. Landing `Stream` in the stdlib has been a mult-year process; and it's incredibly exciting for this to finally happen!

---

r? @KodrAus 
cc/ @rust-lang/wg-async-foundations @rust-lang/libs

yoshuawuyts commented:
oh, I just realized the RFC already sets a feature name: `async_stream`. Updating the PR to gate on that instead.

taiki-e commented on library/core/src/stream/mod.rs:
```diff
@@ -0,0 +1,209 @@
+//! you'll notice something here: we never called anything on our vector to
+//! produce a stream. What gives?
+//!
+//! There's a trait in the standard library for converting something into an
+//! stream: [`IntoStream`]. This trait has one method, [`into_stream`],
+//! which converts the thing implementing [`IntoStream`] into a stream.
+//!
+//! Unlike `std::iter::IntoIterator`, `IntoStream` does not have compiler
+//! support yet. This means that automatic conversions like with `for` loops
+//! doesn't occur yet, and `into_stream` will always have to be called manually.
+//!
+//! [`IntoStream`]: trait.IntoStream.html
+//! [`into_stream`]: trait.IntoStream.html#tymethod.into_stream
+//!
+//! # Adapters
+//!
+//! Functions which take an [`Stream`] and return another [`Stream`] are
+//! often called 'stream adapters', as they are a form of the 'adapter
+//! pattern'.
+//!
+//! Common stream adapters include [`map`], [`take`], and [`filter`].
+//! For more, see their documentation.
+//!
+//! [`map`]: trait.Stream.html#method.map
+//! [`take`]: trait.Stream.html#method.take
+//! [`filter`]: trait.Stream.html#method.filter
+//!
```
```suggestion
```

taiki-e commented on library/core/src/stream/mod.rs:
```diff
@@ -0,0 +1,209 @@
+//!
+//! # Laziness
+//!
+//! Streams (and stream [adapters](#adapters)) are *lazy*. This means that
```
```suggestion
//! Streams are *lazy*. This means that
```

taiki-e commented on library/core/src/stream/mod.rs:
````diff
@@ -0,0 +1,209 @@
+//! example of `while let`:
+//!
+//! ```no_run
+//! # fn main() -> std::io::Result<()> { async_std::task::block_on(async {
+//! #
+//! # use async_std::prelude::*;
+//! # use async_std::stream;
+//! let mut values = stream::repeat(1u8).take(5);
````
you need to remove async-std dep

yoshuawuyts replied:
oops good catch; that's an artifact of copying over parts of the async-std docs. Fixing!

yoshuawuyts replied:
oh this won't even work since we don't have `stream::repeat` yet. We'll need to re-add this in a future PR once we add `stream::repeat`

taiki-e commented on library/core/src/stream/stream/next.rs:
```diff
@@ -0,0 +1,33 @@
+    }
+}
+
+#[unstable(feature = "async_stream", issue = "79024")]
+impl<S: Unpin + ?Sized> Unpin for Next<'_, S> {}
+
```
This can be removed, as `&'a mut S` is always `Unpin`.
```suggestion
```

yoshuawuyts replied:
Ah good catch; not only is this correct -- I also forgot to copy over the impl of `Stream for &mut S`. Fixed both!

taiki-e commented on library/core/src/stream/stream/mod.rs:
```diff
@@ -0,0 +1,97 @@
+        (0, None)
+    }
+
+    /// Advances the stream and returns the next value.
+    ///
+    /// Returns [`None`] when iteration is finished. Individual stream
+    /// implementations may choose to resume iteration, and so calling `next()`
+    /// again may or may not eventually start returning [`Some(Item)`] again at some
+    /// point.
```
I recommend mentioning about the returned type is future, not value.

taiki-e commented on library/core/src/stream/stream/mod.rs:
```diff
@@ -0,0 +1,97 @@
+    {
+        Next::new(self)
+    }
+}
```
```suggestion
}

impl<S: ?Sized + Stream + Unpin> Stream for &mut S {
    type Item = S::Item;

    fn poll_next(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<Self::Item>> {
        S::poll_next(Pin::new(&mut **self), cx)
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        (**self).size_hint()
    }
}

impl<P> Stream for Pin<P>
where
    P: DerefMut + Unpin,
    P::Target: Stream,
{
    type Item = <P::Target as Stream>::Item;

    fn poll_next(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<Self::Item>> {
        self.get_mut().as_mut().poll_next(cx)
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        (**self).size_hint()
    }
}
```

taiki-e replied:
The following two impls are also needed:

`alloc::boxed`
```rust
impl<S: ?Sized + Stream + Unpin> Stream for Box<S> {
    type Item = S::Item;

    fn poll_next(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<Self::Item>> {
        Pin::new(&mut **self).poll_next(cx)
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        (**self).size_hint()
    }
}
```
`std::panic`
```rust
impl<S: Stream> Stream for AssertUnwindSafe<S> {
    type Item = S::Item;

    fn poll_next(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<S::Item>> {
        unsafe { self.map_unchecked_mut(|x| &mut x.0) }.poll_next(cx)
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        self.0.size_hint()
    }
}
```

yoshuawuyts commented:
Updated with @taiki-e's feedback!

taiki-e commented on library/core/src/stream/mod.rs:
````diff
@@ -0,0 +1,158 @@
+//! using it. The compiler will warn us about this kind of behavior:
+//!
+//! ```text
+//! warning: unused result that must be used: streams are lazy and
+//! do nothing unless consumed
````
```suggestion
//! warning: unused result that must be used: streams do nothing unless polled
```

taiki-e commented:
@yoshuawuyts: I guess you missed https://github.com/rust-lang/rust/pull/79023#discussion_r523141096.

nikomatsakis commented:
Technically this should probably be reviewed by someone in the @rust-lang/compiler area, since we're supposed to be managing the actual implementation. I'm going to put

r? @tmandry

taiki-e commented on library/std/src/lib.rs:
```diff
@@ -495,6 +496,9 @@ pub mod task {
 #[stable(feature = "futures_api", since = "1.36.0")]
 pub mod future;
 
+#[unstable(feature = "async_stream", issue = "79024")]
+pub mod stream;
```
Why this is not just a reexport of `core::stream`?

yoshuawuyts replied:
I anticipated `std`-specific extension might arise, and they did! -- this now contains the `AssertUnwindSafe` implementation for `Stream`, which can only be implemented as part of `std` because that's where it's defined.

yoshuawuyts replied:
oh actually; you pointed out they should be implemented on the other submodules -- I think that's indeed better and that would indeed enable this to just be a re-export. Good point!

yoshuawuyts replied:
okay, updated to just make it a re-export!

nagisa commented on library/core/src/stream/mod.rs:
```diff
@@ -0,0 +1,157 @@
+//!
+//! Calling `next().await` this way gets repetitive. Rust has a construct which
+//! can call `next()` on your stream, until it reaches `None`. Let's go over
+//! that next.
```
This says we'll be going over the construct that calls next, but appears to not come back to it? Am I reading this wrong?

yoshuawuyts replied:
You're not reading that wrong; this is indeed incorrectly adapted from `std::iter`

yoshuawuyts replied:
Removed that. Thanks!

bluetech commented on library/core/src/stream/stream/mod.rs:
```diff
@@ -0,0 +1,129 @@
+
+    /// Advances the stream and returns a future which yields the next value.
+    ///
+    /// Returns [`None`] when iteration is finished. Individual stream
```
The word "returns/returning" is slightly confusing - it refers to the future value but can also be construed to refer to the function return value. The sentence above uses the term "yields".

yoshuawuyts commented:
Updated with @bluetech and @nagisa's feedback!

Keruspe commented on library/core/src/stream/stream/mod.rs:
```diff
@@ -0,0 +1,129 @@
+impl<P> Stream for Pin<P>
+where
+    P: DerefMut + Unpin,
+    P::Target: Stream,
```
Nit: shouldn't this be `<P as DerefMut>::Target`?

taiki-e replied:
Both are valid, but I prefer the shorter one.

FWIW, you also can write as follows:
```rust
impl<P> Stream for Pin<P>
where
    P: Unpin + DerefMut<Target: Stream>,
```

guswynn commented on library/alloc/src/boxed.rs:
```diff
@@ -1431,3 +1432,16 @@ impl<F: ?Sized + Future + Unpin, A: AllocRef> Future for Box<F, A> {
+    type Item = S::Item;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Pin::new(&mut **self).poll_next(cx)
```
why the 2 *'s here? the first DerefMut's the Pin, which actually deref's the box: https://doc.rust-lang.org/src/core/pin.rs.html#785-789, the second is extraneous I believe (https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=528d5daa795f4c38210697a6919e67f1), should we match how the future parallel does this: https://doc.rust-lang.org/src/alloc/boxed.rs.html#1162-1168?

I know this is a tiny detail, but I want to make sure I understand if the 2 *'s are necessary

taiki-e replied:
`Pin::new(&mut **self)` returns `Pin<&mut S>`, `Pin::new(&mut *self)` returns `Pin<&mut Box<S>>` (same type as `self`).
I'm not sure why `Future`'s one is actually work, but 2 * should be more correct. (actually, `impl Future for &mut F` uses that.)

guswynn replied:
ohhhh self is a `Pin<&mut Box<S>` so the first * deref's the &mut. autoderef might be doing something funky but you are right, 2 *'s is more explicit and better here

yoshuawuyts commented:
Ping @tmandry; if you have time, a review on this would be great!

sfackler commented:
LGTM, but @nikomatsakis did you want to wait for @tmandry to get a chance to review as well?

tmandry commented on library/core/src/stream/stream/mod.rs:
```diff
@@ -0,0 +1,129 @@
+    /// `poll_next` calls.
+    ///
+    /// - `Poll::Ready(None)` means that the stream has terminated, and
+    /// `poll_next` should not be invoked again.
```
Given that iterators [may choose to resume iteration](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#tymethod.next), I think maybe we should relax this constraint?

taiki-e replied:
Calling `poll_next` after `Poll::Ready(None)` is returned can result in a panic, as described in the "panics" section. (This is especially true when [implementing streams based on generators](https://github.com/taiki-e/futures-async-stream/blob/f516f3a2a77182e99ece456f6e3a5ab0c18c115a/src/lib.rs#L346-L362).)
I think it's okay to create a stream implementation that doesn't end with `Poll::Ready(None)`, but it is preferable to assume all streams end with `Poll::Ready(None)`, except in the rare case where you know the concrete type and it is guaranteed that stream doesn't end with `Poll::Ready(None)`.

taiki-e replied:
...So I think relaxing this constraint would have little benefit to the user and would only confuse it.
(Is there a real use case that benefits from the relaxation of this constraint?)

Nemo157 replied:
(I personally believe that this is exactly the same thing as specified in `Iterator::next` docs, there is no constraint on what an iterator does once it is "finished" so it is perfectly valid for it to panic if polled again; this is just much more explicit about it.)

tmandry replied:
Hmm, if what @Nemo157 is true then I think we should update the Iterator docs to be more explicit. They don't rule out that an iterator may panic, but also don't establish it as a normal thing you should assume in the general case.

@taiki-e My worry was that people would assume that the Iterator contract around this would carry over to Stream (which is a very reasonable assumption IMO). But now it's unclear to me what the contract actually is.

ratmice replied:
I find it a bit odd the contention that its valid for `Iterator::next` to panic, but also return `Some` after it has returned `None`,
given that [FusedIterator](https://doc.rust-lang.org/std/iter/trait.FusedIterator.html)
> Calling next on a fused iterator that has returned None once is guaranteed to return None again.

And anywhere you panic you *could* just as easily return `None`.
There are cases where I have used Non-fused iterators which return `Some`, after returning `None`, It is weird that in the above interpretation panic'ing would be allowed for a non-fused iterator, but not for a fused one, since it is guaranteed to return `None`.

jyn514 replied:
@ratmice FusedIterator is a separate trait from Iterator, so both are true at once: you can have an iterator that returns Some, None, Some, it just won't be a FusedIterator.

I agree though that it's better to return None than to panic - I guess the issue here is there's no equivalent of None for the Stream to return? Can't it just return `Stream::Ready(None)`?

ratmice replied:
@jyn514 Yeah, my point was largely that if you `panic!` on next instead of implementing `FusedIterator`, the `Some`, `None`, `Some` allowance on `Iterator` becomes fairly broken, in these cases it *should* implement `FusedIterator` which does preclude `panic!`.

I have an implementation of a Non-Fused DelayedIterator which implements None while count > 0, and only then starts iterating.  It's largely useless given `slice::windows` not sure if something like that would ever be useful for stream?

I guess my question largely is whether we should try and keep the `Fused`/NonFused distinction intact.  As it seems to me this `stream` is behaving more like `Fused` with a panic instead of repeating None.

tmandry replied:
Part of @taiki-e's point was that generators panic if you resume them after they return. So adapting from generator to Stream would have to involve storing another bit of information (whether it has returned) if panics aren't allowed.

But the generator objects already have that information (it's how they know to panic), and we could find some way to expose it to the Stream adaptor code, if we wanted.

Nemo157 replied:
> @jyn514 Yeah, my point was largely that if you `panic!` on next instead of implementing `FusedIterator`, the `Some`, `None`, `Some` allowance on `Iterator` becomes fairly broken, in these cases it _should_ implement `FusedIterator` which does preclude `panic!`.

As I see it there are three major allowed `Iterator` and `Stream` implementations/requirements:

 1. A `FusedIterator` (and probably `FusedStream` in the future, though [`futures::stream::FusedStream`](https://docs.rs/futures/0.3.8/futures/stream/trait.FusedStream.html) works off a different system where it pre-checks whether it should be polled at all) which guarantees any amount of `Some` followed by an infinite number of `None`.
 2. A "resumeable" iterator which guarantees a mix of `Some` and `None`, and documents under what conditions it switches between them.
 3. A "single-use" iterator which guarantees any amount of `Some`, followed by a single `None`, and then doesn't guarantee its behaviour afterwards.

Distinguishing between the second and third variants is only a matter of documentation, and the majority of the time all that's needed is the third variant, resuming an iterator seems to be a very uncommon thing, especially in a context where you have a generic `I: Iterator` since you don't know the exact conditions under which it switches.

> Part of @taiki-e's point was that generators panic if you resume them after they return. So adapting from generator to Stream would have to involve storing another bit of information (whether it has returned) if panics aren't allowed.

And this will be relevant for iterators in the future when we have generator-backed `Iterator` implementations.

yoshuawuyts replied:
It sounds like we have consensus that `Iterator` and `Stream` are allowed to return items after they've yielded `None` once. But in practice they rarely do, and when created through generators they never will.

I'm inclined to omit the `panic` section for now, and make sure the text from `Stream::next` matches the text from `Iterator::next`. Then in a follow-up PR add a `# panic` heading to both which clarifies that these methods may panic. We can add this as a concern to the tracking issue to make sure this is clarified and consistent between `Iterator` and `Stream` before stabilization.

Does that sound like an acceptable outcome?

taiki-e replied:
> I'm inclined to omit the `panic` section for now, and make sure the text from `Stream::next` matches the text from `Iterator::next`. Then in a follow-up PR add a `# panic` heading to both which clarifies that these methods may panic. We can add this as a concern to the tracking issue to make sure this is clarified and consistent between `Iterator` and `Stream` before stabilization.

I'm in favor of adding a panic section to Iterator, but I don't think it's necessary to remove the panic section from Stream in this patch. The current panic section in this patch is consistent with `Future` and explains exactly what can happen.

yoshuawuyts replied:
Ensuring the documentation of panic behavior on `Stream` and `Iterator` is consistent should be something we resolve before stabilization. I've added an entry for this to the tracking issue: https://github.com/rust-lang/rust/issues/79024.

Going to mark this thread as resolved as this is something which can be addressed in follow-up PRs (and likely requires updates to the `Iterator` docs too).

jyn514 commented on library/core/src/stream/mod.rs:
```diff
@@ -0,0 +1,158 @@
+//! but they are default methods, built on top of [`poll_next`], and so you get
+//! them for free.
+//!
+//! [`Poll`]: ../task/enum.Poll.html
+//! [`Stream`]: trait.Stream.html
+//! [`poll_next`]: trait.Stream.html#method.poll_next
+//! [`next`]: trait.Stream.html#method.next
+//! [`Option`]: ../option/enum.Option.html
```
Yes, all of these are in `core` so you can use them here.
```suggestion
//! [`Poll`]: super::task::Poll
//! [`poll_next`]: Stream::poll_next
//! [`next`]: Stream::next
```

sfackler commented on library/core/src/stream/stream/mod.rs:
```diff
@@ -0,0 +1,129 @@
+    /// [`Some(Item)`] again at some point.
+    ///
+    /// [`Some(Item)`]: Some
+    fn next(&mut self) -> Next<'_, Self>
```
@carllerche notes that this needs a `Self: Sized` bound to ensure that `Stream` is object safe.

seanmonstar replied:
A simple thing I've done before to make sure a trait is object safe is add this to the code next to the trait:

```rust
fn _is_object_safe(_s: &dyn Stream<Item=()>) {}
```

yoshuawuyts replied:
`Stream::next` has been removed from the RFC as per https://github.com/rust-lang/rfcs/pull/2996#issuecomment-757386206. This PR will be updated in accordance.

Darksonn commented on library/core/src/stream/stream/mod.rs:
```diff
@@ -0,0 +1,129 @@
+    /// [`Some(Item)`] again at some point.
+    ///
+    /// [`Some(Item)`]: Some
+    fn next(&mut self) -> Next<'_, Self>
+    where
+        Self: Unpin,
+    {
```
This makes the stream not object safe because it mentions `Self` in the return type.
```suggestion
    fn next(&mut self) -> Next<'_, Self>
    where
        Self: Sized + Unpin,
    {
```
Can we add a test that makes sure that the trait is object safe?

Darksonn replied:
Additionally, please be aware that this can cause problems with calling `next` on a `Pin<Box<dyn Stream>>`. We ran in to the same issue in https://github.com/tokio-rs/tokio/pull/3291 when trying to remove `AsyncReadExt` and `AsyncWriteExt` from Tokio.

For example, to call `.next()` on an `&mut dyn Stream`, you would have to do `(&mut the_stream_ref).next()` to use the impl of `Stream` for a `&mut dyn Stream`. This also requires separate impls of `Stream` for `&mut dyn Stream + Send` and `&mut dyn Stream + Send + Sync` and so on to be able to call the `next()` method at all on those kinds of trait objects.

Darksonn replied:
Here is a playground with a few variants: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ddef891c714e4b244c749a1129e54c22).

taiki-e replied:
@carllerche @Darksonn Good catch! It's not difficult to restore object safety, but not being able to directly call `.next` on `&mut dyn Stream` is bad...

yoshuawuyts replied:
`Stream::next` has been removed from the RFC as per https://github.com/rust-lang/rfcs/pull/2996#issuecomment-757386206. This PR will be updated to reflect that.

camelid commented on library/core/src/stream/mod.rs:
```diff
@@ -0,0 +1,156 @@
+//! let x = counter.next().await.unwrap();
+//! println!("{}", x);
+//! #
+//! }
```
I think this is missing a `#`:

```suggestion
//! # }
```

yoshuawuyts replied:
Ah yeah, good point. However now that we've removed the `next` method, this too has had to go unfortunately.

olanod replied:
Oh I missed when the `next` got removed, I was kind of happy about it since it gives the most basic functionality without an extra dependency.

camelid commented on library/core/src/stream/mod.rs:
```diff
@@ -0,0 +1,156 @@
+//! until you call [`next`]. This is sometimes a source of confusion when
+//! creating a stream solely for its side effects.
+//!
+//! This will not print any values, as we only created a stream, rather than
```
I think it's a little unclear what "This" means. Maybe merge this with the previous paragraph and say something along the lines of "Creating a stream will not print any values, as we only constructed it and didn't use it it."?

yoshuawuyts replied:
Good catch. I copied this from the `Iterator` docs, and think I left out the code example. I've opted to remove the sentence all together for now.

camelid commented on library/core/src/stream/mod.rs:
```diff
@@ -0,0 +1,156 @@
+//! method which is used when implementing a `Stream`, and the [`next`] method
+//! which is used when consuming a stream. Consumers of `Stream` only need to
+//! consider [`next`], which when called, returns a future which yields
+//! [`Option`]`<Item>`.
```
Maybe link to `Stream::Item`?

yoshuawuyts replied:
Oh I like that; done!

camelid reviewed (commented):
A few documentation nits. Great docs by the way :)

yoshuawuyts commented:
`Stream::next` has temporarily been removed to reflect the changes in the RFC; I've added an entry to the tracking issue to ensure we resolve the dynamic dispatch issues before stabilization.

I've also implemented @camelid's feedback, and resolved the other remaining comments on this PR. This should be ready for review by the libs team.

camelid commented on library/core/src/stream/mod.rs:
```diff
@@ -0,0 +1,128 @@
+//! method which is used when implementing a `Stream`, and a (to-be implemented)
+//! `next` method which is used when consuming a stream. Consumers of `Stream`
+//! only need to consider `next`, which when called, returns a future which
+//! yields [`Option`]`<Item>`.
```
Hmm, I think you're missing brackets here:

```suggestion
//! yields [`Option`][`<Item>`].
```

yoshuawuyts replied:
good suggestion; thanks!

camelid commented on library/core/src/stream/mod.rs:
````diff
@@ -0,0 +1,128 @@
+//! ```
+//!
+//! Unlike `Iterator`, `Stream` makes a distinction between the [`poll_next`]
+//! method which is used when implementing a `Stream`, and a (to-be implemented)
````
nit: Putting the hyphen between "to" and "be" but not "be" and "implemented" looks a little funny to me. I think it would be better as none of them hyphenated or all of them hyphenated:

```suggestion
//! method which is used when implementing a `Stream`, and a (to be implemented)
```

```suggestion
//! method which is used when implementing a `Stream`, and a (to-be-implemented)
```

Personally I think the second option looks better than the first, but I think either is good enough.

yoshuawuyts replied:
good one; thanks!

yoshuawuyts commented:
Updated with @camelid's feedback on the docs formatting!

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1750080802) [(plain)](https://github.com/rust-lang/rust/commit/fb7f9aad74caf2a8df83a506637b7dcf507d8768/checks/1750080802/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.15s
Copying stage0 std from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
 Documenting core v0.0.0 (/checkout/library/core)
error: unresolved link to `<Item>`
  --> library/core/src/stream/mod.rs:46:25
   |
46 | //! yields [`Option`][][`<Item>`](Stream::Item).
   |                         ^^^^^^^^ missing type for generic parameters
   |
   = note: `-D broken-intra-doc-links` implied by `-D warnings`
error: aborting due to previous error

error: could not document `core`


Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2018 --crate-type lib --crate-name core library/core/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc --error-format=json --json=diagnostic-rendered-ansi --markdown-css rust.css --markdown-no-toc -Z unstable-options --resource-suffix 1.51.0 --index-page /checkout/src/doc/index.md -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/release/deps --cfg=bootstrap -Dwarnings -Winvalid_codeblock_attributes --crate-version '1.51.0-nightly
  (d33b05f87
  2021-01-22)'` (exit code: 1)


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "core" "--" "--markdown-css" "rust.css" "--markdown-no-toc" "-Z" "unstable-options" "--resource-suffix" "1.51.0" "--index-page" "/checkout/src/doc/index.md"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap doc --stage 0 library/std
Build completed unsuccessfully in 0:00:09
```

</details>

KodrAus commented:
Looks like we're good to go here!

@bors r+

bors commented:
:pushpin: Commit a1b11321fb2d6ce00af9c8957c98df76432b1b78 has been approved by `KodrAus`

<!-- @bors r=KodrAus a1b11321fb2d6ce00af9c8957c98df76432b1b78 -->
<!-- homu: {"type":"Approved","sha":"a1b11321fb2d6ce00af9c8957c98df76432b1b78","approver":"KodrAus"} -->
